src/ILLink.RoslynAnalyzer/LocalDataFlow/LocalDataFlowVisitor.cs

```diff
@@ -0,0 +1,186 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using ILLink.Shared;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace ILLink.RoslynAnalyzer
+{
+	// Visitor which tracks the values of locals in a block. It provides extension points that get called
+	// whenever a value that comes from a tracked local reference flows into one of the following:
+	// - field
+	// - parameter
+	// - method return
+	public abstract class LocalDataFlowVisitor<TValue, TValueLattice> : OperationVisitor<LocalState<TValue>, TValue>,
+		ITransfer<BlockWrapper, LocalState<TValue>, LocalStateLattice<TValue, TValueLattice>>
+		where TValue : IEquatable<TValue>
+		where TValueLattice : ILattice<TValue>
+	{
+		protected readonly LocalStateLattice<TValue, TValueLattice> LocalStateLattice;
+
+		protected readonly OperationBlockAnalysisContext Context;
+
+		protected TValue TopValue => LocalStateLattice.Lattice.ValueLattice.Top;
```

Comment:
Why not fold the LocalStateLattice into this type?

Reply:
Addressed in https://github.com/dotnet/linker/pull/2360#discussion_r748501628